### ISSUE TITLE
Wwhen I try to run on Windows 7 x64, it requires libcairo-2.dll

### ISSUE THREAD

**Issue by AlexeyAB** [sat8/YoloOCLInference#2]
@sat8 Hi, thank you for the great job of porting the Yolo to OpenCL.
I successfully compiled YoloOCLInference on Windows 7 x64 MSVS 2015, but when I try to run, it requires `libcairo-2.dll`
I see that you added all ready-to-use binary library files, but there isn't `libcairo-2.dll`: https://github.com/sat8/YoloOCLInference/tree/master/3rdparty/cairo/lib

**Comment by sat8** [comment]
@AlexeyAB  thnx.

Regarding the error,did u miss step 5 in https://github.com/sat8/YoloOCLInference#dependencies by any chance? Otherwise, ca utry to add the path of gtkmm into your environment path variable? (On my desktop win7 x64, the path for gtkmm is c:\gtkmm64\bin ). The concerned file is definitely in there...

**Comment by AlexeyAB** [comment]
@sat8 Thank you.

Yes, now it works. I use: Ge Froce GTX 970  (GM204) 4 GB GPU-RAM

* but if i run: `YoloOCLInference.exe -video "test2.mp4" -display 0 -save 1` then it processes only first 48 frames and creates only 48 output images (every time processed different number of frames 13, 48, 44, ...), where test2.mp4 has length 1:18 (2262 frames = 78 sec x 29 FPS)

* also not all frames processed if I run: `YoloOCLInference.exe -video "test2.mp4" -display 0 -save 0`
* not all frames processed if I run: `YoloOCLInference.exe -video "test2.mp4" -display 1 -save 0` and I leave the Space-button pressed to avoid freez

![image](https://user-images.githubusercontent.com/4096485/31905017-828ee726-b835-11e7-84d7-9b0ffeb7dae1.png)


![image](https://user-images.githubusercontent.com/4096485/31905006-79176be6-b835-11e7-897c-1f1cf3bd918e.png)

**Comment by sat8** [comment]
@AlexeyAB the code is using ffmpeg for decoding the video & then wraps the YUV frame into OpenCV mat RGB object. Sounds like ffmpeg is reaching a EOF (end of file/stream) signal when it is trying to read packets. But "different number of frames" 13, 48, 44 is not expected. Any chance you can share the test file so I can reproduce it n fix if possible?

**Comment by AlexeyAB** [comment]
@sat8 It happens for any of 6 video mp4 files that I tried on Windows. For example, for this: https://drive.google.com/open?id=0BwRgzHpNbsWBbGZGNkNiaGo2V1k

**Comment by sat8** [comment]
@AlexeyAB ok, thnx for this. Just reading through the code, I found a place where the problem could arise. I will test and fix it hopefully by end of tomorrow..

**Comment by sat8** [comment]
@AlexeyAB Can you update to the latest code and try once more? I have verified the fix and here is the output video generated https://drive.google.com/open?id=0B2aU6UdMlrQbZTBZNk4tbGUtY2M

**Comment by sivagnanamn** [comment]
@sat8  Same error here even with the latest commit. And "cudart64_90.dll" missing.

![image](https://user-images.githubusercontent.com/14053778/32001941-2fb452e6-b9d6-11e7-95df-d36b46b78100.png)
![image](https://user-images.githubusercontent.com/14053778/32001996-52b961a0-b9d6-11e7-8da7-af008874ccf3.png)

**Comment by sat8** [comment]
hmm.

@sivagnanamn I am hoping yours is a 64 bit PC?
am just going to paste my comment from up above. can u check if it helps.

_did u miss step 5 in https://github.com/sat8/YoloOCLInference#dependencies by any chance? Otherwise, can utry to add the path of gtkmm into your environment path variable? (On my desktop win7 x64, the path for gtkmm is c:\gtkmm64\bin ). The concerned file is definitely in there..._

regarding cudart64_90.dll, do u have cuda 9.0 installed? If it is, then try adding the path of CUDA to system path environment variable (you will need to restart cmd window for the path to take effect). My PC has CUDA here **C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\bin**

**Comment by sivagnanamn** [comment]
@sat8  Yes, I'm using an 64 bit machine. I had Gtkmm, adding the path solved the libcairo2.dll issue.

I'm trying this on an AMD GPU with AMD APP SDK installed and configured. How to resolve the cudart64_90.dll issue? Is CUDA mandatory ?

**Comment by sat8** [comment]
@sivagnanamn Ok, understood. I have a GTX1080Ti and OpenCV build local to my PC has created the CUDA runtime dependency (although the code is not using CUDA). I will need to recompile OpenCV without CUDA and push my changes this evening.

**Comment by AlexeyAB** [comment]
@sat8 I updated code from the last commit, but I still get an error:

> Exception at 0x000007FEC1EF61B3 (opencv_core330.dll) in YoloOCLInference.exe: 0xC0000005: access violation at address 0x000000000BA3A040.

**Comment by AlexeyAB** [comment]
@sat8 This is definitely a data race when simultaneous accessing to `m_WorkingImage` from different threads.

An exception occurs here: https://github.com/sat8/YoloOCLInference/blob/1ddb974ef23c985d81c82330130658ebac7b6efc/src/VideoProcessing/YoloOCLDNN.cpp#L1644

----

Used command line: `YoloOCLInference.exe -video "res.avi" -display 0 -save 0`

**Comment by sat8** [comment]
@sivagnanamn I have been able to recompile OpenCV 3.3.0 without CUDA but my dev PC has 3 OpenCL SDKs installed, one each for Intel, AMD & NVIDIA GPUs and it's kind of laborious to uninstall and re-install one SDK at a time.

Would you mind downloading the zip folder shared [here](https://drive.google.com/open?id=0B2aU6UdMlrQbelQ5U3dmaUczd2M) and simply rename your existing **opencv** folder under **3rdparty** folder and copy the unzipped folder to **3rdparty**. (make sure it's just 3rdparty/opencv/...
but not 3rdparty/opencv/opencv/..). Once this is done, delete your build directory and follow the process to build for windows https://github.com/sat8/YoloOCLInference#compilation-in-windows. 
If these binaries work for you, I will commit them to the repo as a permanent fix.

@AlexeyAB thnx for the pointers. I see that there could indeed be a problem. I shall get back on this after some experimentation at my end.

**Comment by sivagnanamn** [comment]
@sat8  Cool, not a problem. Let me check it out later today :+1:

**Comment by sivagnanamn** [comment]
@sat8  I downloaded your opencv build and replaced the old one. Now getting the  below error:

![image](https://user-images.githubusercontent.com/14053778/32052426-2fb147a0-ba93-11e7-9ea5-132139240ae2.png)

`Unhandled exception at 0x0000026054D9CA60 in YoloOCLInference.exe: 0xC0000005: Access violation executing location 0x0000026054D9CA60. occurred`

**Comment by CheekyMonkey** [comment]
FYI
I had the same issues as reported by Alex and Siva

After replacing the opencv folder following Sat's guidance, no now longer face the issues

Windows 10 (64)
NVIDIA GT 1030

\YoloOCLInference\build\Release>YoloOCLInference.exe -input pedestrians.jpg -display 1 -save 0 -det_threshold 0.2 -nms_overlap 0.45

![screenshot 2017-10-26 13 43 55](https://user-images.githubusercontent.com/537672/32053674-eefd491e-ba53-11e7-8070-b2d7cc1c19d6.png)

**Comment by sat8** [comment]
@sivagnanamn It looks like the CUDA dependencies are not there anymore with the OpenCV binaries, which is good. The error could now be occurring during the device initialization stage etc., Can you check if the device is picked up in OpenCL tab of http://www.ozone3d.net/gpu_caps_viewer/.

Meanwhile I am going to add some logging to help identify the sequence of events with a fix for multi threading issue that @AlexeyAB reported.

**Comment by sat8** [comment]
@CheekyMonkey The old OpenCV binaries were compiled with CUDA 9.0 support. In your case, perhaps you have CUDA8.0 installed and hence the errors? Anyway, I have now removed the CUDA dependency in OpenCV binaries and pushed my changes. 

Just to summarize, 
Presently, CMake will pickup any OpenCL library installed on the build PC (it does not differentiate b/n CUDA 8.0/9.0 etc.,) But OpenCV binaries were built with CUDA 9.0 support so the dll's won't work if 9.0 version of CUDA is not installed. Hopefully my latest commit https://github.com/sat8/YoloOCLInference/commit/f6891f270826da5766b1d2cc27874366ce46bd27 will fix this OpenCV <--> CUDA compatibility issue.

**Comment by AlexeyAB** [comment]
@sat8 The simplest way to solve it, replace this line: https://github.com/sat8/YoloOCLInference/blob/1ddb974ef23c985d81c82330130658ebac7b6efc/src/VideoProcessing/YoloOCLDNN.cpp#L1274

To this: `m_CurrentImage.copyTo(m_WorkingImage);`

But probably that an error still possible on the first frame.

**Comment by sat8** [comment]
@AlexeyAB I have been caught up in two thoughts on the multi-threading issue. (mine is i7 6770 PC and I don't see it but I acknowledge the code is problematic). there are 3 potential fixes
1. As you indicated, make a copy rather than clone, but the risk is still there
2. Use a double buffer for m_WorkingImage. Each thread uses alternating buffer indices.
3. Use a STL queue between the decoder thread and inference thread. (this means locking a mutex, unlocking etc which could add some latency but this is the safest of all)

I am implementing (3) presently as an experiment. In the long term, I intend to get the decoder output directly into the OpenCL buffer to optimize on inference speed but it requires quite a few changes in the current implementation. (future work). I shall update here once (3) is done so you can test on your setup.

**Comment by sivagnanamn** [comment]
@sat8  Yup, GPU caps viewer can detect my OpenCL installation.

![image](https://user-images.githubusercontent.com/14053778/32055360-8de201da-ba9c-11e7-84fa-ac5ca1a68314.png)

**Comment by sat8** [comment]
@AlexeyAB , @sivagnanamn @CheekyMonkey 

Here is an update on my latest commit https://github.com/sat8/YoloOCLInference/commit/40a752a158645abb2a87afaa7dc9b8396303957d

1. A fix for multi-threading issue - a STL queue & a mutex ensure thread safety
2. When **-save 1** is set, a H.264 AVI file will be written to disk into output folder.
3. A  logger object dumps the sequence of events into YoloOCLInference.log file - (Work in progress)

@AlexeyAB , I hope this commit will fix the issue you have come across.
@sivagnanamn Can you run VS2015 project (YoloOCLInference.sln) in **RelWithDebInfo** mode as then the IDE break into the place where the crash was occurring so we can investigate further?

Any feedback is much appreciated.

**Comment by CheekyMonkey** [comment]
@sat8 , nice one :)

The new commit works for me (40a752a)

I do see what I believe may be an issue though

For this command:
YoloOCLInference.exe -video testvideo1.mp4 -display 1 -save 1 -det_threshold 0.2 -nms_overlap 0.45

It seems to require that I close the video output window for every single frame that is being processed in order for object detection processing to continue to the next frame

This behaviour also occurred with the previous build

Am I running the command in the correct way?

Build log included below for reference

G:\projects>git clone https://github.com/sat8/YoloOCLInference
Cloning into 'YoloOCLInference'...
remote: Counting objects: 1349, done.
remote: Total 1349 (delta 0), reused 0 (delta 0), pack-reused 1349
Receiving objects: 100% (1349/1349), 158.21 MiB | 1.99 MiB/s, done.
Resolving deltas: 100% (597/597), done.
Checking connectivity... done.
Checking out files: 100% (523/523), done.

G:\projects>cd YoloOCLInference

G:\projects\YoloOCLInference>mkdir build

G:\projects\YoloOCLInference>cd build

G:\projects\YoloOCLInference\build>cmake .. -G "Visual Studio 14 2015 Win64"
-- The C compiler identification is MSVC 19.0.24210.0
-- The CXX compiler identification is MSVC 19.0.24210.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - found
-- Found OpenCL: G:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.0/lib/x64/OpenCL.lib (found version "1.2")
-- Configuring done
-- Generating done
-- Build files have been written to: G:/projects/YoloOCLInference/build

G:\projects\YoloOCLInference\build>

**Comment by sat8** [comment]
@CheekyMonkey thnx.

The display option **-display 1** is not required if you are happy with the detection performance. This feature can be considered a **backdoor** visualization feature incase the user needs to fine tune **det_threshold** & **-nms_overlap** to improve upon mis-detections and/or false-positives. Did I make sense here?

It's not mandatory to provide **-display** argument in case you are not interested in checking the quality of detections for every frame. For instance, in this case, I would just do 
  
   YoloOCLInference.exe -video testvideo1.mp4  -save 1 -det_threshold 0.2 -nms_overlap 0.45

This will save the inference output of the complete video as H.264 encoded AVI to **InferenceOutput.avi** in  output folder. The ultimate goal with YoloOCLInference is to build a C++ API/library wherein a callback would inform the client code about the results (bounding boxes with classes/pre-drawn JPEG/H.264 encoded video packets/other forms etc.,). This way, the client code will handle the results (push to a database, save to files etc.,) depending on the application setting.

**Comment by sivagnanamn** [comment]
@sat8  The code breaks at `clGetDeviceIDs`

![image](https://user-images.githubusercontent.com/14053778/32143810-c250366e-bcf2-11e7-9f8b-a998f9a77429.png)

**Comment by sat8** [comment]
@sivagnanamn I have been able to reproduce this issue with AMD devices. The latest commit https://github.com/sat8/YoloOCLInference/commit/42de1668e9594a6cb240d29354e13163cc140c77 fixes this issue. Couple of notes
- A new command line argument **-gpu_type** is supported from this commit going forward. This allows the user to set a platform type (AMD, NVIDIA, ARM) if multiple OpenCL devices are available on the target machine.
- Detection output can be saved either as JPEG images or H.264 AVI file (set the command line argument **-save** to 1 or 2 respectively)

Can you pls verify if the issue is resolved at your end? thnx

**Comment by sivagnanamn** [comment]
@sat8  Debug output after latest commit.

![image](https://user-images.githubusercontent.com/14053778/32374491-5377f63a-c0e1-11e7-960d-54e8748a0060.png)

**Comment by sat8** [comment]
@sivagnanamn My RX580 seems to not complain. Is your AMD gpu integrated in to the CPU (not yet tested at my end)?
Also, were you able to run some example apps successfully from the official [AMD APP SDK](http://developer.amd.com/amd-accelerated-parallel-processing-app-sdk/) without issues?

**Comment by sivagnanamn** [comment]
@sat8  Mine is not an integrated GPU card. I'm able to run other OpenCL codes from Visual studio.  Usually I don't use Windows for my experiments, this is the first time I'm  using it. So I'm not able to help you by debugging it by myself, sorry for that.
